src/hardware/siglent-sds/protocol.c

```diff
@@ -36,6 +36,137 @@
 #include "scpi.h"
 #include "protocol.h"
 
+/** Wait for 500 ms while reading SCPI data */ 
+#define WAIT_DATA_TIMEOUT	500000
+
+/**
+ * Wait and read data from SCPI device.
+ *
+ * @param scpi Previously initialised SCPI device structure.
+ * @param buf Buffer to store result.
+ * @param maxlen Maximum number of bytes to read.
+ *
+ * @return Number of bytes read, or SR_ERR upon failure.
+ */
+static int siglent_scpi_wait_read_data(struct sr_scpi_dev_inst *scpi,
+			char *buf, int maxlen)
+{
+	int64_t start = g_get_monotonic_time();
+	int64_t elapsed;
+	int ret;
+	do {
+		ret = sr_scpi_read_data(scpi, buf, maxlen);
+		elapsed = g_get_monotonic_time() - start;
+	} while ((ret == 0) && (elapsed < WAIT_DATA_TIMEOUT));
+	return ret;
+}
+
+/**
+ * Get the number of points of the current wave and save it to devc->num_samples.
+ *
+ * @param sdi the device instance.
+ * @param devc the device context.
+ *
+ * @return SR_ERR upon failure.
+ */
+static int siglent_get_wave_points(struct sr_dev_inst *sdi,struct dev_context *devc)
```

Comment:
Nitpick: I don't fancy the output argument devc. When reading the function usage, it is not immediately clear what member fields are updated. Either num_samples could be passed directly (and devc would never enter the function), or the function could be renamed to reflect it updated num_samples field only (`siglent_update_num_samples`).